### ISSUE TITLE
Evaluation/Metrics approach

### ISSUE THREAD

**Issue by jasonkriss** [pytorch/ignite#78]
This has been discussed a bit in other issues, but I wanted to make a dedicated issue for us to discuss this as I think it's very important we get this right.

## Background
Some previous discussions [here](https://github.com/pytorch/ignite/pull/54#issuecomment-363231705) and [here](https://github.com/pytorch/ignite/issues/72).

Throughout, I am going to use the motivating example of training a supervised model where you periodically want to compute some metrics against a validation set.

## Current Setup
Currently, in order to accomplish this, you need to do the following:

1. create an `Evaluator`
2. register the `Evaluate` handler to run the `Evaluator` on the validation set and store the predictions in the history
3. add another event handler to actually use this history to compute the metrics you care about
4. log/plot these metrics however you choose

In code, this looks something like this:

```python
model = ...
validation_loader = ...
trainer = ...
evaluator = create_supervised_evaluator(model, cuda=True)
trainer.add_event_handler(Events.EPOCH_COMPLETED, Evaluate(evaluator, validation_loader, epoch_interval=1))
@trainer.on(Events.EPOCH_COMPLETED)
def log(engine):
    print(engine.current_epoch, categorical_accuracy(evaluator.history))
```

### Pros
- keeps library code cleanly separated with minimal implicit dependencies
- user doesn't have to write much code

### Cons
- can be confusing what happens where
  - we have both an `Evaluator` and an `Evaluate` and yet neither one computes any sort of metrics
- there are a lot of ways this could go wrong
  - you have to understand the contract between what gets stored in the `Evaluator`'s history and the metrics functions
  - you have to make sure you attach the `Evaluate` handler and any logging handlers to the same event

## Goals
Evaluating a model is gonna be something that (essentially) everyone does so I think we need to have a good story here. Imo, we should make the supervised case be super easy while still making it possible for non-supervised cases. That being said, I think we want to accomplish this without removing flexibility and without adding a ton of code.

## Ideas
Working backward from what I would like the api to be, it might be nice if you could just do something like this:

```python
model = ...
validation_loader = ...
trainer = ...
@trainer.on(Events.EPOCH_COMPLETED)
def evaluate(engine):
  results = evaluate(model, {'acc': categorical_accuracy})
  # do something with those results
```

It'd be even nicer if I could do something like this:
```python
model = ...
validation_loader = ...
trainer = ...
trainer.add_event_handler(Events.EPOCH_COMPLETED, Evaluate(model, {'acc': categorical_accuracy}))
```

But without making assumptions about how users want to plot/log their evaluation results, this isn't possible. 

What do you all think? Anything here you take issue with? Any ideas on how we can best accomplish this? Do we need to make plotting/logging part of this discussion as well?

**Comment by alykhantejani** [comment]
I think this needs to be addressed asap and I think you're correct with the approach. We should hash out what we want the API to look like for evaluation, we want to make it super easy for the most common case - evaluation using pre-canned metrics for a supervised task. But also make more complex things possible.

I will try to have a think about what I think the API should look like and post back here, if others could do the same hopefully we will converge on something nice soon

**Comment by alykhantejani** [comment]
So I came up with something pretty similar to @jasonkriss's:

```python
def run(train_batch_size, val_batch_size, epochs, lr, momentum, log_interval, logger):
    train_loader, val_loader = get_data_loaders(train_batch_size, val_batch_size)
    model = Net()
    optimizer = SGD(model.parameters(), lr=lr, momentum=momentum)
    trainer = create_supervised_trainer(model, optimizer, nn.NLLLoss())


    @trainer.on(Events.ITERATION_COMPLETED)
    def log_training_loss(trainer):
        if trainer.current_iteration % log_interval == 0:
            log_simple_moving_average(trainer, window_size=10, logger=logger)


    @trainer.on(Events.EPOCH_COMPLETED)
    def log_validation_results(trainer):
        evaluator = Evaluator(model, metrics=[categorical_accuracy, nn.NLLLoss()], mean=[True, True])
        avg_accuracy, avg_loss = evaluator.run(val_loader)
        logger("Validation Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}"
               .format(trainer.current_epoch, avg_accuracy, avg_loss))

    trainer.run(train_loader, max_epochs=epochs)
```

**Comment by jasonkriss** [comment]
My personal preference would be for the reduction behavior to be completely encapsulated in the metrics themselves. I think this allows the greatest flexibility and prevents us from having to worry about things like how to average accuracy across batches (if the batches are not all the same size, then we cant average them correctly). 

I also would love to keep Evaluator as an instance of `Engine` without giving it more responsibility.

So I think that opens up two other options, building on your ideas...

## Option 1: Evaluate Handler
- assumes the existence of a state object (ala #80)
- we'd need to have one for supervised, one for unsupervised etc
- the evaluator still exists but it is essentially private in this case

Here is a rough sketch:

```python
class Evaluate(object):
  def __init__(model, dataloader, metrics=[]):
    self._model = model
    self._dataloader = dataloader
    self._metrics = metrics
    self._evaluator = create_supervised_evaluator(model)

  def __call__(state):
    history = self._evaluator.run(self._dataloader).history
    state.metrics = [metric(history) for metric in self._metrics]
```

```python
model = ...
validation_loader = ...
trainer = ...
trainer.add_event_handler(Events.EPOCH_COMPLETED, Evaluate(model, validation_loader, metrics=[categorical_accuracy, nn.NLLLoss()]))
@trainer.on(Events.EPOCH_COMPLETED)
def log(state):
  acc, loss = state.metrics
  print(acc, loss)
```

or, if we don't like passing around the metrics on the state object, we could have it be a function like this:

```python
model = ...
validation_loader = ...
trainer = ...
@trainer.on(Events.EPOCH_COMPLETED)
def eval(state):
  acc, loss = evaluate(model, validation_loader, metrics=[categorical_accuracy, nn.NLLLoss()])
  print(acc, loss)
```

## Option 2: Stateful metrics
- similar to tensorflow (tf.metrics) and torchnet's meters
- could be implemented similar to the timer in #74 
  - meaning they are objects that attach themselves as handlers to multiple events on the evaluator

Very rough sketch...

```python
class Metric(object):
  def __init__(self, name):
    self._name = name
    self.reset()

  def update(self, y_pred, y):
    raise NotImplementedError

  def value(self):
    raise NotImplementedError

  def reset(self):
    raise NotImplementedError

  def iteration_completed(state):
    self.update(*state.history[-1])

  def completed(state):
    state.metrics[self._name] = self.value()
    self.reset()

  def attach(self, evaluator):
    evaluator.add_event_handler(Events.ITERATION_COMPLETED, self.iteration_completed)
    evaluator.add_event_handler(Events.ITERATION_COMPLETED, self.completed)

class CategoricalAccuracy(Metric):
  def reset(self):
    self._total = 0
    self._count = 0

  def update(self, y_pre, y):
    indices = torch.max(y_pred, 1)[1]
    correct = torch.eq(indices, y)
    self._total += torch.sum(correct)
    self._count + = correct.shape[0]

  def value(state):
    return self._count / self._total
```

```python
def create_supervised_evaluator(model, metrics=[], cuda=False):
  ...
  evaluator = Evaluator(_inference)
  for metric in metrics:
    metric.attach(evaluator)
  return evaluator
```

```python
model = ...
validation_loader = ...
trainer = ...
evaluator = create_supervised_evaluator(model, metrics=[CategoricalAccuracy('acc')])
@trainer.on(Events.EPOCH_COMPLETED)
def eval(_):
  state = evaluator.run(validation_loader)
  print(state.metrics['acc'])
```

Obviously still very rough, but wanted to throw it out there. What do you all think?

**Comment by alykhantejani** [comment]
With the evaluate handler, the problem is that it stores the full history and then does the metric calculation at the end. This means that for something like imagenet classification you are storing validation dataset size * 1000 numbers when you really only want 1000 numbers (correct/incorrect) and ultimately a mean. 

However, I really like the stateful metrics approach, would you mind opening a PR to start this effort? 

I think we should also outline the difference between state + history and also what should be passed to handlers (but I will comment more on this in #80)

**Comment by jasonkriss** [comment]
@alykhantejani That's a great point. I'll start putting together a PR for the stateful approach.

With that in mind, how would you feel about removing `History` and the `handlers.logging`. Instead of history, we can have `state.output` contain the most recent iterations output. That's all that will be needed for metrics. This way we don't store a bunch of data unless its truly needed. The rest of the methods on history can be introduced as `metric`s if we need them. 

I think it'd be nice to simplify the library as much as possible. Having both history/logging and the metrics/state approach might lead to confusion among users who aren't sure which approach to use.

What do you think?

**Comment by alykhantejani** [comment]
@jasonkriss sounds like a good idea, and like we're on the same page - see [my recent comment here](https://github.com/pytorch/ignite/issues/80#issuecomment-366318519).

**Comment by henrique** [comment]
Hey guys, I'm trying to understand what happened to the metrics history. Is it stored anywhere currently? or what is the best way for storing all validation scores during training?
Thanks for the good work!

**Comment by vfdev-5** [comment]
Hey @henrique , yes, now metrics history is not stored by engine's state. I would say the best way to keep it available is to log it : print, tensorboardx, write into file or any long-term living variable to plot, ...).
Take a look at our [notebook examples](https://github.com/pytorch/ignite/tree/master/examples/notebooks):
- log to TensorboardX: EfficientNet_Cifar100_finetuning or CycleGAN
- log to a variable to plot with matplotlib: FashionMNIST

HTH
